### PR TITLE
Don't display full docs in sys.doc by default

### PR DESCRIPTION
**Pull request by cachedout**
This is a huge amount of data and really unnecessary. Instead, display a list of functions and show how to query individual functions for full documentation.

This PR is up for discussion.

cc: @thatch45, @terminalmage @rallytime

### CODE REVIEW

File: salt/modules/sysmod.py [hunk 1]
```diff
@@ -37,8 +37,8 @@ def __virtual__():
 
 def doc(*args):
     '''
-    Return the docstrings for all modules. Optionally, specify a module or a
-    function to narrow the selection.
+    Return docstring for a module. If not module is passed,
```
Comment (terminalmage) [inline salt/modules/sysmod.py]: Should probably be ``no`` instead of ``not``.

File: salt/modules/sysmod.py [hunk 2]
```diff
@@ -65,9 +65,10 @@ def doc(*args):
     '''
     docs = {}
     if not args:
-        for fun in __salt__:
-            docs[fun] = __salt__[fun].__doc__
-        return _strip_rst(docs)
+        funcs = __salt__.keys()
```
Comment (terminalmage) [inline salt/modules/sysmod.py]: Is there a specific reason for the switch to ``.keys()`` here? On Python 3 it is a dict_keys view, so no issue there, but on Python 2 it allocates a new list in memory. IIRC, simply iterating over a dict (like it was doing before) doesn't allocate a new list.

**Review by terminalmage — COMMENTED** [review]
Looks great, just a couple questions.

**Comment by thatch45** [comment]
I am fine with this approach, this is a great deal more data than it when when I originally wrote it!
@whiteinge does the enterprise UI use sys.doc at all?

**Comment by whiteinge** [comment]
Salt Enterprise does not use this but other UIs do (cc @Lothiraldan) and I've seen this used in the wild in scripts. Because this data is very often used programmatically we can't change this behavior without the usual **deprecation path** and a good replacement function that can fetch all docs from all modules in one call.

IMO, the [docs Runner](https://docs.saltstack.com/en/latest/ref/runners/all/salt.runners.doc.html) is the best replacement. It was written, in part, for this use-case and it has been on my wish-list to change the `docs.execution` function to gather docstrings from modules locally on the Master without having to hit minions.

There's no reason we have to fetch that information by querying minions since most of it is duplicate data. As part of that change I would like it to use the `raw()` loader and return docstrings for modules even if they wouldn't normally load on that system. This doesn't address modules with the same `__virtualname__`, of course, but neither does the current `sys.doc` so maybe that's ok. Thoughts welcome.

**Comment by whiteinge** [comment]
I misspoke. Salt Enterprise does use this function, although indirectly.

To clarify: I think this is a **good change**. We don't need all the duplicate information from each minion individually. But we do need a single function call that can return all docs for all modules.

**Comment by damon-atkins** [comment]
> There's no reason we have to fetch that information by querying minions since most of it is duplicate data. 
@whiteinge   pkg is an example where there are differences based on the OS.

**Comment by whiteinge** [comment]
@damon-atkins Sorry, I split my thoughts across two comments there. See the end of the first one:

> This doesn't address modules with the same __virtualname__, of course, but neither does the current sys.doc so maybe that's ok. Thoughts welcome.

The current `sys.doc` (targeted to `*`) is last-wins (I think). Sure you can target individual minions for specific docs but that's something of an accident. You still have to know a lot about that minion to know what docs you're looking at. We definitely have the opportunity with this change to expose this information in a better way. Perhaps we include docs for each file name whenever there is a `__virtualname__` collision. E.g.:

```json
{"pkg": {
    "yumpkg": "...docs...",
    "aptpkg": "...docs...",
}
```

Though that feels a tad kludgy.

**Comment by whiteinge** [comment]
Oh. You were quoting from that exact comment. Sorry. Time for a nap I think.

**Comment by cachedout** [comment]
@whiteinge The big concern for me at least is the size of the data inbound to the master and not necessarily the display *of* that data. Even if we consolidate that data prior to display, we're still looking at 12GB of data inbound to a master over 10K minions for a single `sys.doc`.

The secondary problem is that delivering that data, even from a single minion which has a large number of functions available, is that the return is so large that it actually exceeds the maximum size of a singe event that we've enforced across the event bus.

We can increase that but it's starting to feel a little ridiculous, to be honest. We can (and should) also consider starting to allow for multi-part events or returns made up of multiple events or however we want to end up structuring this but that might be a totally different discussion.

Anyhow, just some additional context for the discussion. :]

**Comment by thatch45** [comment]
I am starting to think that we should change this to be a runner, the master can get all the docs in all the modules now. Admittedly, I would have made it a runner originally if I had a yet developed a master minion concept

**Comment by cachedout** [comment]
Yes, that is what I am thinking too. Though, that does make the stand-alone salt-call end of things a little tricky.

**Comment by thatch45** [comment]
sys.doc is a call the intercepts the remote execution routine to temper the return, we could keep the minion side sys.doc and just redirect the master side to the runner. Sloppy yes, but it is not
too unlike how it already works....

**Comment by whiteinge** [comment]
If we added all modules it looks like the full docs return might be in the ~1.5-2 MB range. I wouldn't bat an eye at that as output from a CLI command but that does feel large to call an "event". What is the current size threshold for an event payload?

**Comment by cachedout** [comment]
@whiteinge Currently 1MB

**Comment by whiteinge** [comment]
Hm. Correct me if I'm wrong, but with event-based Runners that would make returning all docs a non-starter without adjusting that number, right?

🤔

Thinking this through from the perspective of, "What's the ideal for an autocompleting UI?" I suppose the ideal would be to fetch just the list of functions and then to fetch the _individual_ docs on-demand after that so that we only actually load the data that we actually use. (Ideally from the Runner we're discussing, but one that could dole out individual docs.) But our Runners are not even close to fast enough to do that for a UI right now.

Spitballing:

Say we made the docs Runner only return individual docs (like in this PR), salt-api could hide that slowness by calling the docs Runner for each individual module and caching the result in, say, a `/docs` endpoint to be able to quickly dole out HTTP responses for individual docs. (If someone was calling the Python API instead that would be on them to build.)

That doesn't strike me as an awful addition, off the top of my head, and it's well within the purview of salt-api. An argument could be made that this need is specific to salt-api, after all.

Are there other good solutions? Does the multi-part event thing need to happen eventually anyway? It strikes me that a large, programmatically generated highstate return wouldn't have too much difficulty hitting that 1 MB limit. (Not a common occurrence, of course, but a show-stopper for anyone that did hit it.)

**Comment by cachedout** [comment]
Closing this until we can get back to this.

**Pull request closed without merging**